sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -562,6 +572,8 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
   protected override def nullSafeEval(value: Any): Any = {
     if (set.contains(value)) {
       true
+    } else if (isNaN(value)) {
```

Comment:
shall we check nan before calling `set.contains(value)`?

Reply:
> shall we check nan before calling `set.contains(value)`?

Notice this, since it's `nullSafeEval `,  `null` has been checked

Reply:
is it a bug? this means we return null if `value` is null, no matter `hasNull` is true or false.

Reply:
Codegen seems fine as we will do a null check at the very end.

Reply:
> is it a bug? this means we return null if `value` is null, no matter `hasNull` is true or false.

`In` have same behavior,  
```
  override def eval(input: InternalRow): Any = {
    val evaluatedValue = value.eval(input)
    if (evaluatedValue == null) {
      null
    } else {
      var hasNull = false
      list.foreach { e =>
        val v = e.eval(input)
        if (v == null) {
          hasNull = true
        } else if (ordering.equiv(v, evaluatedValue)) {
          return true
        }
      }
      if (hasNull) {
        null
      } else {
        false
      }
    }
  }
```